In [ ]:
array_of_dictionaries = [{
    "name": "Budi",
    "age": 10,
    "test_scores": [100.0, 98.0, 89.0]
},
{
    "name": "Charlie",
    "age": 12,
    "test_scores": [90.0, 100.0]
}]

def get_students_averages(dictionary_list):
  for element in dictionary_list:
      name = dictionary_list[0]['name']
      name2 = dictionary_list[1]['name']
      age = dictionary_list[0]['age']
      age2 = dictionary_list[1]['age']
      avg = round(sum(dictionary_list[0]['test_scores'])/len(dictionary_list[0]['test_scores']), 2)
      avg2 = round(sum(dictionary_list[1]['test_scores'])/len(dictionary_list[1]['test_scores']), 2)

      print("Nomor 1")
      print('Pelajar ' + name + ' dengan umur ' + str(age) + ' mendapatkan nilai rata-rata ' + str(avg))
      print('Pelajar ' + name2 + ' dengan umur ' + str(age2) + ' mendapatkan nilai rata-rata ' + str(avg))
      break

get_students_averages(array_of_dictionaries)

def get_students_min_score(dictionary_list):
  for element in dictionary_list:
    name = dictionary_list[0]['name']
    name2 = dictionary_list[1]['name']
    age = dictionary_list[0]['age']
    age2 = dictionary_list[1]['age']
    minus = min(dictionary_list[0]['test_scores'])
    minus2 = min(dictionary_list[1]['test_scores'])

    print("Nomor 2")
    print('Pelajar ' + name + ' dengan umur ' + str(age) + ' mendapatkan nilai terkecil ' + str(minus))
    print('Pelajar ' + name2 + ' dengan umur ' + str(age2) + ' mendapatkan nilai terkecil ' + str(minus2))
    break

get_students_min_score(array_of_dictionaries)

def get_students_max_score(dictionary_list):
  for element in dictionary_list:
    name = dictionary_list[0]['name']
    name2 = dictionary_list[1]['name']
    age = dictionary_list[0]['age']
    age2 = dictionary_list[1]['age']
    bigest = max(dictionary_list[0]['test_scores'])
    bigest2 = max(dictionary_list[1]['test_scores'])

    print("Nomor 3")
    print('Pelajar ' + name + ' dengan umur ' + str(age) + ' mendapatkan nilai terbesar ' + str(bigest))
    print('Pelajar ' + name2 + ' dengan umur ' + str(age2) + ' mendapatkan nilai terbesar ' + str(bigest2))
    break
    
get_students_max_score(array_of_dictionaries)

def get_max_score_by_student(dictionary_list, student_name):
  for element in dictionary_list:
      for attribute in element.keys():
          if element['name'] == student_name:
            age = element['age']
            bigest = max(element['test_scores'])

            print("Nomor 4")
            print('Pelajar ' + student_name + ' dengan umur ' + str(age) + ' mendapatkan nilai terbesar ' + str(bigest))
            break

get_max_score_by_student(array_of_dictionaries, 'Charlie')

def add_student(dictionary_list, student_dictionary):
  dictionary_list.append(student_dictionary)
  return dictionary_list

updated_dictionary = add_student(array_of_dictionaries, {"name": "Doddy", "age": 13, "test_scores": [100.0, 100.0, 100.0]})
print("Nomor 6")
print("Dictionary setelah di update =" ,updated_dictionary)

Nomor 1
Pelajar Budi dengan umur 10 mendapatkan nilai rata-rata 95.67
Pelajar Charlie dengan umur 12 mendapatkan nilai rata-rata 95.67
Nomor 2
Pelajar Budi dengan umur 10 mendapatkan nilai terkecil 89.0
Pelajar Charlie dengan umur 12 mendapatkan nilai terkecil 90.0
Nomor 3
Pelajar Budi dengan umur 10 mendapatkan nilai terbesar 100.0
Pelajar Charlie dengan umur 12 mendapatkan nilai terbesar 100.0
Nomor 4
Pelajar Charlie dengan umur 12 mendapatkan nilai terbesar 100.0
Nomor 6
Dictionary setelah di update = [{'name': 'Budi', 'age': 10, 'test_scores': [100.0, 98.0, 89.0]}, {'name': 'Charlie', 'age': 12, 'test_scores': [90.0, 100.0]}, {'name': 'Doddy', 'age': 13, 'test_scores': [100.0, 100.0, 100.0]}]


1. Membuat function get_students_averages() untuk mendapatkan nilai rata-rata ujian
untuk masing-masing student dari array array_of_dictionaries. Function menerima 
parameter:
◦ dictionary_list (array of dictionaries)
2. Lalu, function akan menjalankan function print() yang menghasilkan output seperti
berikut:
Pelajar Budi dengan umur 10 mendapatkan nilai rata-rata 95.67
Pelajar Charlie dengan umur 12 mendapatkan nilai rata-rata 95.0
2. Membuat function get_students_min_score() untuk mendapatkan nilai ujian terkecil
untuk masing-masing student dari array array_of_dictionaries. Function menerima 
parameter:
◦ dictionary_list (array of dictionaries)
3. Lalu, function akan menjalankan function print() yang menghasilkan output seperti
berikut:
Pelajar Budi dengan umur 10 mendapatkan nilai terkecil 89.0 Pelajar Charlie
dengan umur 12 mendapatkan nilai terkecil 90.0
4. Membuat function get_students_max_score() untuk mendapatkan nilai ujian
terbesar untuk masing-masing student dari array array_of_dictionaries. Function
menerima 1 parameter:
◦ dictionary_list (array of dictionaries)
5. Lalu, function akan menjalankan function print() yang menghasilkan output seperti
berikut:
Pelajar Budi dengan umur 10 mendapatkan nilai terbesar 89.0 Pelajar Charlie
dengan umur 12 mendapatkan nilai terbesar 100.0
6. Membuat function get_max_score_by_student() untuk mendapatkan nilai ujian
terbesar untuk suatu student dari array array_of_dictionaries. Function menerima 2
parameter:
◦ dictionary_list (array of dictionaries)
◦ student_name (nama pelajar yang ingin dicari nilai terbesarnya)
7. Lalu, function akan menjalankan function print() yang menghasilkan output seperti
berikut ketika function get_max_score_by_student(dictionary_sample, "Charlie")
dijalankan:
Pelajar Charlie dengan umur 12 mendapatkan nilai terbesar 100.0
8. Membuat function get_all_time_max_avg_score() untuk mendapatkan nilai rata-rata
ujian terbesar dari semua pelajar. Function menerima 1 parameter:
◦ dictionary_list (array of dictionaries)
9. Lalu, function akan menjalankan function print() yang menghasilkan output seperti
berikut:
Pelajar Budi dengan umur 10 mendapatkan nilai rata-rata terbesar, yaitu 95.67
10. Membuat function add_student() untuk menambahkan seorang pelajar ke array
dictionary yang sudah ada. Function menerima 2 parameter:
◦ dictionary_list (array of dictionaries)
◦ student_dictionary (dictionary student yang akan dimasukan ke array)
11. Lalu, function akan memberikan return value dictionary baru yang sudah
ditambahkan dengan student_dictionary yang diberikan di parameter.
Sehingga, ketika kode ini dijalankan
updated_dictionary = add_student(array_of_dictionaries, { "name": "Doddy", "age":
13, "test_scores": [100.0, 100.0, 100.0] })
print(updated_dictionary)
 Maka, sistem akan mengeluarkan isi list yang sudah ditambahkan dengan
dictionary baru.